Feature: ignore actors

Laurin-W opened the pull request:
Addresses https://github.com/assemblee-virtuelle/activitypods/issues/65

I wrote unit tests that seem to work fine (even though they are quite slow). So the backend seems to work fine now.

However, with the front-end, I haven't figured out yet, why the `apods:ignoredContacts` collection is not pulled from the server. Maybe you could have a look at this @srosset81? :)

srosset81 commented:
Thanks, I will look at your problem ASAP.

srosset81 commented:
Hi @Laurin-W,

First, you didn't use the correct predicate for the collection: it's `apods:ignoredContacts` and not `apods:ignored` (maybe it was part of your tests...):
https://github.com/assemblee-virtuelle/activitypods/pull/69/files#diff-29cfce18778af1053850ced258a7260f97d348055923a861762e3b4391bb13afR12

Second, and this is probably the real problem, the `apods:ignoredContacts` predicate was not on [ActivityPods JSON-LD context](https://activitypods.org/context.json) (which, for now, is used for all requests) so when fetching the actor, the data were formatted like an object with an `id` property:

![Capture d’écran de 2023-06-07 18-17-46](https://github.com/assemblee-virtuelle/activitypods/assets/17931931/bc5f5381-885c-410a-a78d-5875c2a29e0f)

This is [corrected](https://github.com/assemblee-virtuelle/activitypods/commit/88df3e4c4e1c6470954223be2d565502d05b6476) and deployed on prod, so it should work. At least it does for me, the URI of the ignored contacts are correctly fetched by the `useCollection` hook !

Haven't look at the rest of your PR, but it looks promising ! Thanks.

PS: JSON-LD context is generally a pain to deal with. I hope to make it easier with the planned refactoring (NLNet grant)

Laurin-W commented on boilerplate/services/mixins/notification-filter.js:
```diff
@@ -0,0 +1,25 @@
+      const actor = await this.broker.call('activitypub.actor.get', { actorUri: activity.actor });
+  
+      const isIgnored = await this.broker.call('activitypub.collection.includes', {
+        collectionUri: actor['apods:ignoredContacts'],
```
Created https://github.com/assemblee-virtuelle/semapps/pull/1131 for this

Laurin-W commented:
Alright, thanks for the review! I addressed your feedback.
Note, that I created a PR:  https://github.com/assemblee-virtuelle/semapps/pull/1131 on which this PR depends.

srosset81 commented on boilerplate/services/notification.service.js:
```diff
@@ -1,13 +1,24 @@
-    ? [SingleMailNotificationsService, QueueService(CONFIG.QUEUE_SERVICE_URL)]
-    : [SingleMailNotificationsService],
+    ? [SingleMailNotificationsService, notificationFilter, QueueService(CONFIG.QUEUE_SERVICE_URL)]
+    : [SingleMailNotificationsService, notificationFilter],
```
Unfortunately there is still a problem. In Moleculer the order of the mixins is important. The notificationFilter mixin must be before the SingleMailNotificationsService, otherwise the `filterNotification` method of SingleMailNotificationsService overwrites the `filterNotification` method of the notificationFilter mixin. This is a bit contrary to common sense !

srosset81 commented on tests/contacts.test.js:
```diff
@@ -18,10 +21,11 @@ beforeAll(async () => {
   // Mock notification service
   await broker.createService({
-    mixins: [require('./services/notification.service')],
+    mixins: [notificationFilter, require('./services/notification.service')],
```
Your integration tests passed because you put the mixins on the correct order here :-)

srosset81 commented on boilerplate/services/mixins/MailNotificationFilterMixin.js:
```diff
@@ -0,0 +1,24 @@
+        itemUri: activity.actor,
+      });
+      if (isIgnored) {
+        console.log('ignoring notification for', notification, activity);
```
I think it's useful to keep a log of this. But it's better to use Moleculer logger and to make it more compact: `this.logger.info(`Ignoring notification for activity ${activity.id} sent to ${recipientUri}`)`

srosset81 reviewed (changes requested):
Just one last problem ! And otherwise it works great !

Laurin-W commented:
Alright, thanks for the second review :)

srosset81 commented:
Many thanks. I will deploy this on monday.
